### ISSUE TITLE
add negroni middleware

### ISSUE THREAD

**Issue by earlzo** [elastic/apm-agent-go#665]
**Is your feature request related to a problem? Please describe.**

https://github.com/urfave/negroni

> Negroni is an idiomatic approach to web middleware in Go. It is tiny, non-intrusive, and encourages use of net/http Handlers.

the `modules/apmhttp` is a little bit inconvenient when we have multiple middlewares.

the code using `modules/apmhttp` likes that:

```go
func setupGlobalMiddleware(handler http.Handler) http.Handler {
	return sentryhttp.New(sentryhttp.Options{}).Handle(apmhttp.Wrap(handler))
}
```

Negroni is a good package for `http` compares to web frameworks with middleware support.

**Describe the solution you'd like**

add `module/apmnegroni`

```go
func setupGlobalMiddleware(handler http.Handler) http.Handler {
	n := negroni.New()
	n.Use(sentrynegroni.New(sentrynegroni.Options{}))
	n.Use(apmnegroni.Middleware())
	n.UseHandler(handler)
	return n
}
```

**Additional context**

I'd like to implement this feature if you accept this feature request

**Comment by axw** [comment]
@earlzo I think it's reasonable to add this. Before you go ahead and implement this, could you please explain how you plan to set the transaction names?

`apmhttp` currently uses the URL path as the transaction name, which means that transactions will not be aggregated properly. Because Negroni is not aware of route patterns, I guess this middleware would have to do the same thing. I'm worried that this will lead to a bad experience in the APM UI.

Which router(s) do you use along with Negroni? Do you just use `net/http.ServeMux`?

**Comment by earlzo** [comment]
We are using `go-swagger` for our api service and it uses https://github.com/naoina/denco as it's router. 

So,

Do we just use net/http.ServeMux?

No.

Is there a better solution for the `http` package?

Maybe we can provide an optional `Routers` interface, the agent use it to look up the router for a request.

**Comment by axw** [comment]
> Maybe we can provide an optional Routers interface, the agent use it to look up the router path for a request.

When using `apmhttp.Wrap`, you can pass in a function to extract the transaction name from the request, using the option [apmhttp.WithServerRequestName](https://godoc.org/go.elastic.co/apm/module/apmhttp#WithServerRequestName). If you make it possible to pass options into `apmnegroni.Middleware()`, then you could, for example, fetch the route by using `denco.Router.Lookup`.

Feel free to go ahead with implementing this feature. Thank you!

**Comment by axw** [comment]
Closed by #671
